radixdlt-core/radixdlt/src/main/java/com/radixdlt/api/core/reconstruction/BerkeleyRecoverableProcessedTxnStore.java

```diff
@@ -0,0 +1,261 @@
+	@Override
+	public void open(DatabaseEnvironment dbEnv) {
+		recoverableTransactionsDatabase = dbEnv.getEnvironment().openDatabase(null, RECOVERABLE_TRANSACTIONS_DB_NAME, new DatabaseConfig()
+			.setAllowCreate(true)
+			.setTransactional(true)
+			.setKeyPrefixing(true)
+			.setBtreeComparator(lexicographicalComparator())
+		);
+
+		accumulatorDatabase = dbEnv.getEnvironment().openDatabase(null, ACCUMULATOR_HASH_DB_NAME, new DatabaseConfig()
+			.setAllowCreate(true)
+			.setTransactional(true)
+			.setKeyPrefixing(true)
+			.setBtreeComparator(lexicographicalComparator())
+		);
+
+		try (var cursor = accumulatorDatabase.openCursor(null, null)) {
+			var key = new DatabaseEntry(Longs.toByteArray(Long.MAX_VALUE));
+			var value = new DatabaseEntry();
+			cursor.getSearchKeyRange(key, value, null);
+			var status = cursor.getPrev(key, value, null);
+			if (status == SUCCESS) {
+				var accumulatorHash = HashCode.fromBytes(value.getData());
+				var stateVersion = Longs.fromByteArray(key.getData());
+				this.accumulatorState = new AccumulatorState(stateVersion, accumulatorHash);
+			} else {
+				this.accumulatorState = new AccumulatorState(0, HashUtils.zero256());
+			}
+		}
+	}
+
+	@Override
+	public void close() {
+		if (recoverableTransactionsDatabase != null) {
+			recoverableTransactionsDatabase.close();
+		}
+
+		if (accumulatorDatabase != null) {
+			accumulatorDatabase.close();
+		}
+	}
+
+	public Optional<HashCode> getAccumulator(long stateVersion) {
+		if (stateVersion == 0) {
+			return Optional.of(HashUtils.zero256());
+		}
+		var key = new DatabaseEntry(Longs.toByteArray(stateVersion));
+		var value = new DatabaseEntry();
+		var result = accumulatorDatabase.get(null, key, value, null);
+		if (result != SUCCESS) {
+			return Optional.empty();
+		}
+		return Optional.of(HashCode.fromBytes(value.getData()));
+	}
+
+	public List<RecoverableProcessedTxn> get(long index, long limit) {
+		try (var cursor = recoverableTransactionsDatabase.openCursor(null, null)) {
+			var iterator = new Iterator<RecoverableProcessedTxn>() {
+				final DatabaseEntry key = new DatabaseEntry(Longs.toByteArray(index));
+				final DatabaseEntry value = new DatabaseEntry();
+				OperationStatus status = cursor.get(key, value, Get.SEARCH, null) != null ? SUCCESS : OperationStatus.NOTFOUND;
+
+				@Override
+				public boolean hasNext() {
+					return status == SUCCESS;
+				}
+
+				@Override
+				public RecoverableProcessedTxn next() {
+					if (status != SUCCESS) {
+						throw new NoSuchElementException();
+					}
+					RecoverableProcessedTxn next;
+					try {
+						next = serialization.fromDson(Compress.uncompress(value.getData()), RecoverableProcessedTxn.class);
+					} catch (IOException e) {
+						throw new IllegalStateException("Failed to deserialize committed transaction.", e);
+					}
+
+					status = cursor.getNext(key, value, null);
+					return next;
+				}
+			};
+			return Streams.stream(iterator)
+				.limit(limit)
+				.onClose(cursor::close)
```

Comment:
what's that for? isn't the cursor already getting closed by the try block?

Reply:
Yup, good catch, this leftovers from previous refactor :)